### PR TITLE
Add Scopes configuration options to MSAL Electron's Public Client Application

### PR DESCRIPTION
**Pull request by technical-boy**
This PR:

- Adds the scaffolding for `PublicClientApplication#acquireToken()` method
- Adds scope configuration options through `AuthenticationParameters` class for `PublicClientApplication#acquireToken()`
- Adds error handling for scopes configuration
- Adds passing tests for scope configuration error handling.

### CODE REVIEW

File: lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts [hunk 1]
```diff
@@ -13,4 +15,35 @@ export class PublicClientApplication extends ClientApplicationBase {
...
+    /**
+     * The acquireToken method uses the Authorization Code
+     * Grant to retrieve an access token from the AAD authorization server,
+     * which can be used to make authenticated calls to an resource server
+     * such as MS Graph.
+     */
+    public acquireToken(request: AuthenticationParameters): string {
+        // Validate and filter scopes
+        this.validateInputScopes(request.scopes);
+        const scope = request.scopes.join(' ').toLowerCase();
```
Comment (jasonnutter) [inline lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts]: Whats the point of this line?

File: lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts [hunk 2]
```diff
@@ -13,4 +15,35 @@ export class PublicClientApplication extends ClientApplicationBase {
...
+        return 'Access Token';
+    }
+
+    private validateInputScopes(scopes: string[]): void {
+        // Check that scopes are present
+        if (!scopes) {
+            throw ClientConfigurationError.createScopesRequiredError(scopes);
+        }
```
[the remark this answers is not in the document]
Comment (jasonnutter) [reply lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts]: And if this is just used in one place, you don't need to factor it out into its own function.

File: lib/msal-electron-proof-of-concept/src/Utils/UriUtils.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        let canonicalUri: string;
+        if (rawUri) {
+            canonicalUri = rawUri.toLowerCase();
+        }
+
+        if (rawUri && !UriUtils.validateSuffix(rawUri, '/')) {
+            canonicalUri += '/';
+        }
+
+        return canonicalUri;
```
[the remark this answers is not in the document]
Comment (technical-boy) [reply lib/msal-electron-proof-of-concept/src/Utils/UriUtils.ts]: Do you mean keep the method and just change the implementation to use `url` inside? Or scrap the method and use the `url` module wherever we need to normalize URIs?

File: lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (technical-boy) [reply lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts]: Great suggestion, thanks! I'll try to remember using assert when I'm just checking for truthiness.

File: lib/msal-electron-proof-of-concept/src/Utils/UriUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (technical-boy) [reply lib/msal-electron-proof-of-concept/src/Utils/UriUtils.ts]: @jasonnutter changed it to use `url` module inside, is that better?

File: lib/msal-electron-proof-of-concept/test/Utils/UriUtils.spec.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import { expect } from 'chai';
+import * as Mocha from 'mocha';
+
+import { TEST_CONFIGURATION, TEST_URIS } from '../testConstants'; 
+import { UriUtils } from '../../src/utils/uriUtils';
+
+describe('uriUtils.ts class', () => {
+    describe('#canonicalizeUri', () => {
+        it('should return undefined if called with null object', () => {
+            expect(UriUtils.canonicalizeUri(null)).to.be.undefined;
```
[the remark this answers is not in the document]
Comment (technical-boy) [reply lib/msal-electron-proof-of-concept/test/Utils/UriUtils.spec.ts]: I've added an error for this case and changed the test so it expects a throw now.

File: lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts (hunk 1 above)
Comment (technical-boy) [reply lib/msal-electron-proof-of-concept/src/AppConfig/PublicClientApplication.ts]: I thought msal-core did it as pre-processing before adding them to the request URL but it turns out that string was actually used as an object key that we won't be using in msal-electron, so I'm removing it.

File: lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts [hunk 5]
```diff
@@ -29,7 +30,10 @@ export default class Main {
         Main.mainWindow.on('closed', Main.onClose);
 
         // Once appplication is ready, configure MSAL Public Client App
-        Main.setupAuth();
+        Main.configureAuthentication();
+
+        // Listen for AcquireToken button call
+        Main.listenForAcquireToken();
```
Comment (DarylThayil) [inline lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts]: should this only be spun up when a token request is sent, or is it okay to always have open?

File: lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts [hunk 6]
```diff
@@ -40,15 +44,28 @@ export default class Main {
                 nodeIntegration: true,
             },
         });
+        this.mainWindow.webContents.openDevTools();
```
Comment (DarylThayil) [inline lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts]: what does this do?

File: lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts (hunk 5 above)
Comment (DarylThayil) [reply lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts]: OH i see nevermind

File: lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts (hunk 6 above)
Comment (technical-boy) [reply lib/msal-electron-proof-of-concept/samples/msal-electron-sample-app/src/Main.ts]: It opens chromium's dev tools on the window programatically (for debugging purposes). I want to keep it on during development. I'll most likely remove this line by the time Auth Code is done. Should I remove it or comment it?

File: lib/msal-electron-proof-of-concept/src/Utils/UriUtils.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jasonnutter) [reply lib/msal-electron-proof-of-concept/src/Utils/UriUtils.ts]: Yeah, although I'm surprised you need to still check for the presence of the trailing slash, I expected that to be normalized by the `url` module.

[the remark this answers is not in the document]
Comment (technical-boy) [reply lib/msal-electron-proof-of-concept/src/Utils/UriUtils.ts]: Yeah, I though the same when you suggested using `url`. I couldn't find a way to configure `.parse` to add trailing slashes if missing.

**Pull request merged**